pkg/controller/controllercmd/builder.go

```diff
@@ -304,6 +305,15 @@ func (b *ControllerBuilder) Run(ctx context.Context, config *unstructured.Unstru
 		return nil
 	}
 
+	infraStatus, err := clusterstatus.GetClusterInfraStatus(ctx, clientConfig)
+	if err != nil {
+		eventRecorder.Warningf("ClusterInfrastructureStatus", "Failed to get cluster infrastructure status because of %s", err)
+		klog.Warningf("unable to get cluster infrastructure status for leader election (falling back to default): %v", err)
+	} else if infraStatus.ControlPlaneTopology == configv1.SingleReplicaTopologyMode {
+		snoLeaderElection := leaderelectionconverter.LeaderElectionSNOConfig(*b.leaderElection)
+		b.leaderElection = &snoLeaderElection
+	}
```

Comment:
Shouldn't that block move to `WithleaderElection` since it's possible that LE is disabled and this code can accidentally turn it on if we're not careful enough.

Reply:
That's a good point, the reason why it's there is because we won't know the topology we're in until runtime, the function currently won't alter the `disable` flag in LE, but to make it more robust I changed the if statement to be `if topology = SingleReplica AND LE is not disabled` as well as some tests to ensure behavior, do think that's a good enough safeguard against accidentally turnning it on?

Reply:
This code is still in a spot that does not have access to the user's intent.  @soltysh's comment looks unaddressed to me.